### DISCUSSION TITLE
Avalonia on Raspberry pi with Framebuffer - stuck

### DISCUSSION THREAD

**Discussion by dinonovak** [AvaloniaUI/Avalonia#9336] [Q&A] [answered]
HI,
I created a very basic sample https://gitlab.com/dn-misc/AvaloniaRaspberryTest
trying to get Avalonia work on Raspberry PI with FrameBuffer.

I am trying to run it on Raspbian Buster
.Net Framework installed is 3.1.2

When running the project with --fbdev flag it is not generating any error but nothing appears or changes on the screen.

When running in GUI, Avalonia app shows up in window.

Any help how to troubleshoot or make it work.

Thanx,
D

**Comment by kekekeks** [comment]
Try to run it from a proper virtual console (the ones, accessible via Ctrl+Alt+F(1-12) with XOrg stopped

**Comment by dinonovak** [comment]
HI,
tried, but no luck,
tried running as root, but it just get's stuck, it does not try to initiate screen, just gets stuck in console.
Any way to log something to pinpoint issue ?

Thanx,
D

**Comment by kekekeks** [comment]
Try to fill framebuffer with garbage using `cat /dev/urandom > /dev/fb0`. You need to do that with an actual virtual console currently on the screen. Once you succeed with that, you should be able to run Avalonia app on the same terminal.

**Comment by dinonovak** [comment]
OK,
I am getting cat: write error: No space left on device
and snowy picture - mainly white background with random color pixels.

I am trying to run avalonia app after this and when run it clears this white showy screen and I get a clear console.

I made a video of test
https://youtu.be/-t_AjlL-Y4E

Regards,
D

**Comment by dinonovak** [comment]
@kekekeks, can you plase check video I atrached to previous post, I am not sure what am I doing wrong, but still not able to run in framebuffer

**Comment by kekekeks** [comment]
Well, I don't have a raspberry device nearby, so I can only recommend running ControlCatalog with debugger attached and investigating what's going on in this file https://github.com/AvaloniaUI/Avalonia/blob/master/src/Linux/Avalonia.LinuxFramebuffer/Output/FbdevOutput.cs

**Comment by dinonovak** [comment]
@kekekeks I am now trying to use DRM as at least I am getting some feedback.
I installed kmscube and it is working ok.

when trying to run with --drm option I am getting following error

```
pi@minipi:~/avalonia $ dotnet exec AvaloniaTest1.dll --drm
Unhandled exception. System.ComponentModel.Win32Exception (0xFFFFFFFE): drmModeAddFb failed
   at Avalonia.LinuxFramebuffer.Output.DrmOutput.GetFbIdForBo(IntPtr bo)
   at Avalonia.LinuxFramebuffer.Output.DrmOutput.Init(DrmCard card, DrmResources resources, DrmConnector connector, DrmModeInfo modeInfo)
   at Avalonia.LinuxFramebuffer.Output.DrmOutput..ctor(String path)
   at LinuxFramebufferPlatformExtensions.StartLinuxDrm[T](T builder, String[] args, String card, Double scaling)
   at AvaloniaTest1.Program.Main(String[] args) in C:\Users\dino\source\repos\AvaloniaTest1\AvaloniaTest1\Program.cs:line 63
Aborted
```

this is where it all breaks in the code (from https://gitlab.com/dn-misc/AvaloniaRaspberryTest/-/blob/master/AvaloniaTest1/Program.cs)

```
            else if (args.Contains("--drm"))
            {
                SilenceConsole();
                return builder.StartLinuxDrm(args, scaling: GetScaling());
            }
```

avalonia updated to 0.9.11

**Comment by kekekeks** [comment]
There is a somewhat relevant patch for enabling DRM output on ARM devices - https://github.com/AvaloniaUI/Avalonia/pull/4049

**Comment by dinonovak** [comment]
@kekekeks I am using Avalonia 0.9.11 nugets, whish are dated from 4 days ago. I can see that "kekekeks merged commit 52701a0 into AvaloniaUI:master  on 31 May" so shoul'd it be already incorporated in these nugets?

**Comment by kekekeks** [comment]
@dinonovak that patch was merged to `master`, not to the 0.9.x maintenance branch

**Comment by dinonovak** [comment]
@kekekeks 
Ok I now configured VS to use nightly nugets, so with --drm option i now am getting following error

```
Unhandled exception. System.EntryPointNotFoundException: Unable to find an entry point named 'sk_refcnt_safe_unref' in shared library 'libSkiaSharp'.
   at SkiaSharp.SkiaApi.sk_refcnt_safe_unref(IntPtr refcnt)
   at SkiaSharp.SKObjectExtensions.SafeUnRef(ISKReferenceCounted obj)
   at SkiaSharp.SKObject.DisposeNative()
   at SkiaSharp.SKNativeObject.Dispose(Boolean disposing)
   at SkiaSharp.SKObject.Dispose(Boolean disposing)
   at SkiaSharp.GRGlInterface.Dispose(Boolean disposing)
   at SkiaSharp.SKNativeObject.DisposeInternal()
   at SkiaSharp.SKNativeObject.Dispose()
   at Avalonia.Skia.GlSkiaGpu..ctor(IWindowingPlatformGlFeature gl, Nullable`1 maxResourceBytes)
   at Avalonia.Skia.PlatformRenderInterface..ctor(ISkiaGpu skiaGpu, Nullable`1 maxResourceBytes)
   at Avalonia.Skia.SkiaPlatform.Initialize(SkiaOptions options)
   at Avalonia.SkiaApplicationExtensions.<>c__0`1.<UseSkia>b__0_0()
   at Avalonia.Controls.AppBuilderBase`1.Setup()
   at Avalonia.Controls.AppBuilderBase`1.SetupWithLifetime(IApplicationLifetime lifetime)
   at LinuxFramebufferPlatformExtensions.StartLinuxDirect[T](T builder, String[] args, IOutputBackend backend)
   at LinuxFramebufferPlatformExtensions.StartLinuxDrm[T](T builder, String[] args, String card, Double scaling)
   at AvaloniaTest1.Program.Main(String[] args) in C:\Users\dino\source\repos\AvaloniaTest1\AvaloniaTest1\Program.cs:line 63
```

These are referenced nugets:
<img width="436" alt="Screenshot 2020-07-08 at 00 17 42" src="https://user-images.githubusercontent.com/18152490/86849870-a0a27580-c0b0-11ea-8dfc-59eb849f7634.png">


This is Program.cs
```
using System;
using System.Threading;
using System.Diagnostics;
using System.Globalization;
using System.Linq;
using Avalonia;
using Avalonia.Controls.ApplicationLifetimes;
//using Avalonia.Logging.Serilog;
using Avalonia.ReactiveUI;
using Avalonia.Dialogs;

namespace AvaloniaTest1
{
    class Program
    {
        // Initialization code. Don't use any Avalonia, third-party APIs or any
        // Avalonia configuration, don't remove; also used by visual designer.
        //public static AppBuilder BuildAvaloniaApp()
        //    => AppBuilder.Configure<App>()
        //        .UsePlatformDetect()
        //        .LogToDebug();


        // SynchronizationContext-reliant code before AppMain is called: things aren't initialized
        // yet and stuff might break.
        //public static void Main(string[] args) => BuildAvaloniaApp()
        //    .StartWithClassicDesktopLifetime(args);


        [STAThread]
        static int Main(string[] args)
        {
            if (args.Contains("--wait-for-attach"))
            {
                Console.WriteLine("Attach debugger and use 'Set next statement'");
                while (true)
                {
                    Thread.Sleep(100);
                    if (Debugger.IsAttached)
                        break;
                }
            }

            var builder = BuildAvaloniaApp();

            double GetScaling()
            {
                var idx = Array.IndexOf(args, "--scaling");
                if (idx != 0 && args.Length > idx + 1 &&
                    double.TryParse(args[idx + 1], NumberStyles.Any, CultureInfo.InvariantCulture, out var scaling))
                    return scaling;
                return 1;
            }
            if (args.Contains("--fbdev"))
            {
                Console.WriteLine(">> Run in Mode: Framebuffer");
                //SilenceConsole();
                return builder.StartLinuxFbDev(args, scaling: GetScaling());
            }
            else if (args.Contains("--drm"))
            {
                SilenceConsole();
                return builder.StartLinuxDrm(args, scaling: GetScaling());
            }
            else
                return builder.StartWithClassicDesktopLifetime(args);
                //return builder.StartLinuxFbDev(args, scaling: GetScaling());
        }

        /// <summary>
        /// This method is needed for IDE previewer infrastructure
        /// </summary>
        public static AppBuilder BuildAvaloniaApp()
            => AppBuilder.Configure<App>()
                .UsePlatformDetect()
                .LogToDebug()
                .With(new X11PlatformOptions
                {
                    EnableMultiTouch = true,
                    UseDBusMenu = true
                })
                .With(new Win32PlatformOptions
                {
                    EnableMultitouch = true,
                    AllowEglInitialization = true
                })
                .UseSkia()
                .UseReactiveUI()
                .UseManagedSystemDialogs();

        static void SilenceConsole()
        {
            new Thread(() =>
            {
                Console.CursorVisible = false;
                while (true)
                    Console.ReadKey(true);
            })
            { IsBackground = true }.Start();
        }
    }
}
```

**Comment by kekekeks** [comment]
SkiaSharp managed/native version mismatch. I think there is currently some work being done on providing an ARM build out of the box for SkiaSharp 2.80. 

I think @ahopper has a pre-built Skia binary that works with the current master

**Comment by Gillibald** [comment]
This has all the info https://github.com/mono/SkiaSharp/pull/1382

**Comment by ahopper** [comment]
@Gillibald do the builds here mono/SkiaSharp#1382 need https://github.com/AvaloniaUI/Avalonia/pull/4222 ? I convinced myself they did.  My builds for the current master can be extracted from the arm release here http://www.ihopper.org/radio/previews.htm you need libHarfBuzzSharp.so and libSkiaSharp.so . Hopefully that will be the last release with custom builds.

**Comment by dinonovak** [comment]
Hi all,
ok, now I am not getting any error at startup.
But still I am just getting blank screen. Is there anything that would need to be done to MainWindow.xaml to make it work under --drm option. With X Windows started app starts and window is displayed.

raspberry pi config.txt has following for framebuffer:
```
# uncomment to force a console size. By default it will be display's size minus
# overscan.
framebuffer_width=1280
framebuffer_height=720

[all]
dtoverlay=vc4-fkms-v3d
```

Here is MainWindow.xaml definition:
```
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:local="clr-namespace:AvaloniaTest1"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="600"
        x:Class="AvaloniaTest1.MainWindow"
        Title="AvaloniaTest1">
  <Panel>
    <local:MainView />
  </Panel>
</Window>
```
Here is the link to app source: https://gitlab.com/dn-misc/AvaloniaRaspberryTest

**Comment by kekekeks** [comment]
Try enabling "Full KMS" mode in raspberry pi options.

**Comment by dinonovak** [comment]
No luck, still blank screen with Full KMS enabled.

**Comment by adoconnection** [comment]
Having exactly the same behavior on Raspberry Pi 3 B+, Debian Stretch 9.8, .NET Core 3.1

**Comment by adoconnection** [comment]
simple NET Core 3.1 console app with this code works no matter whether I run it directly on Pi or over SSH console
so its not some kind of NET stuff.

```cs
using System;
using System.IO;

namespace ConsoleApp33
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.CursorVisible = false;

            var ScreenWidth = 640;
            var ScreenHeight = 480;
            var BytesPerPixel = 2;

            var random = new Random();
            var buffer = new Byte[ScreenWidth * ScreenHeight * BytesPerPixel];

            using (var fb = new FileStream("/dev/fb0", FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite))
            {
                random.NextBytes(buffer);
                fb.Write(buffer, 0, buffer.Length);
                fb.Close();
            }

            Console.WriteLine("Hello World!");
        }
    }
}
```

**Comment by FFugi** [comment]
Hello, @dinonovak I had similar problem like the one you had at the beginning of this topic (running with --fbdev option)
I think I could share my solution here.

I had very similar issue - when running the app with `StartLinuxFbDev` I did not see anything except my console. 
It looks like when running with `StartLinuxFbDev` we don't have `IClassicDesktopStyleApplicationLifetime`, but ` ISingleViewApplicationLifetime`

And in your code you're checking for `IClassicDesktopStyleApplicationLifetime`, you should check for ` ISingleViewApplicationLifetime` and assign your view to its MainView property:
https://gitlab.com/dn-misc/AvaloniaRaspberryTest/-/blob/c0918cc7a712ed2c45ff69225c3c888bbdf36b91/AvaloniaTest1/App.xaml.cs#L16

**Comment by kekekeks** [comment]
We should probably add some default content like "MainView is not set" message or something. It's confusing otherwise

**Comment by adoconnection** [comment, marked as answer]
For anyone struggling with framebuffers. MainView can not be inialized with Window it will result with ```Could not CreateWindow(): IWindowingPlatform is not registered``` Window has to be replaced with UserControl

Not obvious at all

```cs
if (ApplicationLifetime is ISingleViewApplicationLifetime single)
{
    single.MainView = new Window1(); // Could not CreateWindow(): IWindowingPlatform is not registered
}
```

```cs
if (ApplicationLifetime is ISingleViewApplicationLifetime single)
{
    single.MainView = new UserControl1(); // OK
}
```

**Comment by gutblender** [comment]
I have the same issue as @dinonovak though my Win32Exception message says 
`drmModeSetCrtc failed`
when running with `--drm`.

Running with `--fbdev` never displays a GUI. Running with neither shows it, but there's a flicker whenever anything is redrawn, e.g. on mouse-over buttons in the sidebar, etc. I also tried with AvaloniaEdit.Demo with similar results. So I think it's not the fault of ControlCatalog.NetCore.

Just for a proof of concept, I'm running ControlCatalog.NetCore built from master, on a DragonBoard 410c running the Linaro image provided for it on 96boards.org, linaro-sid-alip-dragonboard-410c-933 w/ .NET Core SDK 3.1.405

**Comment by timunie** [comment]
Converted this into Q&A as this is neither a bug nor a feature request. If you disagree please let us know.
